### PR TITLE
DONT MERGE: assign enhancement to replace #49

### PR DESCRIPTION
**Pull request by cherifGsoul**
New PR to replace #49 the tests still break on ```can-util/dom/attr: attributes event```

Maybe something wrong on attr see discussion here https://github.com/canjs/can-util/issues/230

### CODE REVIEW

**Comment by cherifGsoul** [comment]
Using ```deep-assign``` fixes the issue because we need a clone of ```mutation``` object in this line https://github.com/canjs/can-util/blob/master/dom/events/attributes/attributes.js#L43

File: js/assign/assign.js [hunk 1]
```diff
@@ -22,9 +22,20 @@
...
  */
-module.exports = function (d, s) {
-	for (var prop in s) {
-		d[prop] = s[prop];
-	}
-	return d;
-};
+ module.exports = function (target, args) {
+ 	if (target === null) {
+   	throw new TypeError('The target object cannot be undefined or null');
```
Comment (justinbmeyer) [inline js/assign/assign.js:27]: do we need this type check?  Maybe dev-mode only?   Btw === null won't check undefined.

File: js/assign/assign.js [hunk 2]
```diff
@@ -22,9 +22,20 @@
...
+   }
+ 	for (var i = 1; i < arguments.length; i++) {
```
Comment (justinbmeyer) [inline js/assign/assign.js:29]: we should probably cache `arguments.length`.

File: js/assign/assign.js [hunk 3]
```diff
@@ -22,9 +22,20 @@
...
+ 		var supplier = arguments[i];
+
+ 		if (supplier !=null) {
```
Comment (justinbmeyer) [inline js/assign/assign.js:32]: you'll need a jshint ignore for this.  There's a way to do it w/i a single function.  Look for `jshint:` in the codebase.

File: js/assign/assign.js [hunk 4]
```diff
@@ -22,9 +22,20 @@
...
+ 			for (var prop in supplier) {
+ 				if (Object.prototype.hasOwnProperty.call(supplier, prop)) {
```
Comment (justinbmeyer) [inline js/assign/assign.js:34]: `Objet.prototype.hasOwnProperty` should probably be cached so we don't have to access every property each time.

**Comment by justinbmeyer** [comment]
@cherifGsoul Thanks for following up on this.  We probably should create an `assignMany` or a `simpleAssign` to maintain good assign performance.  I remember adding a simple assign to CanJS 2.X b/c jQuery's assign was so slow.

Can you describe what the error in `attr` is?

**Comment by cherifGsoul** [comment]
Thank you @justinbmeyer the [CanJS 2.x](https://github.com/canjs/canjs/blob/2.3-legacy/util/can.js#L45) is the actual behavior of assign in ```can-util``` .

I dont think that the problem is in attr, ```dom/events/attributes/attributes.addEventListener``` stops all the tests without any outcome when my version of ```assign``` tries to assign ```MutationRecord``` object keys-values in this [line](https://github.com/canjs/can-util/blob/master/dom/events/attributes/attributes.js#L43).

If we use simpleAssign (like the old simpleExtend) attributes.addEventListener will be fixed.

File: js/assign/assign.js (hunk 3 above)
Comment (cherifGsoul) [reply js/assign/assign.js:32]: What to be ignore exactly? all 3 lines?

File: js/assign/assign.js (hunk 4 above)
Comment (cherifGsoul) [reply js/assign/assign.js:34]: Here you mean to cache the returning result or:
```
var has = Object.prototype.hasOwnProperty;
...
if (has.call(supplier, prop)) {
...
}
...
```

**Comment by andrejewski** [comment]
Closing as can-util probably will not go 4.0. In 4.0, we will have `can-assign`, `can-reflect.assign`, and the native `Object.assign` which supports variadic arguments.

**Pull request closed without merging**